felix/deltatracker/delta_tracker.go

```diff
@@ -0,0 +1,335 @@
+	logrus.Debug("DeleteAll")
+	c.IterDesired(func(k K, v V) {
+		c.DeleteDesired(k)
+	})
+}
+
+// IterDesired iterates over the desired KVs.
+func (c *DeltaTracker[K, V]) IterDesired(f func(k K, v V)) {
+	for k, v := range c.desiredUpdates {
+		f(k, v)
+	}
+	for k, v := range c.inDataplaneAndDesired {
+		if _, ok := c.desiredUpdates[k]; ok {
+			continue
+		}
+		f(k, v)
+	}
+}
+
+// ReplaceDataplaneCacheFromIter clears the dataplane cache and replaces its contents with the KVs returned
+// by the iterator.  The pending update and deletion tracking is updated accordingly.
+func (c *DeltaTracker[K, V]) ReplaceDataplaneCacheFromIter(iter func(func(k K, v V)) error) error {
+	logrus.Debug("Loading cache of dataplane state.")
+
+	// To reduce occupancy and to do the update in one pass, we use the old maps as scratch space.
+	// As we iterate over the new values, we add them to the newXXX maps and delete them from the old.
+	// Then, at the end, the oldXXX maps will contain only the values that are now missing from the
+	// dataplane, we then handle those at the bottom.
+	oldInDPDesired := c.inDataplaneAndDesired
+	newInDPDesired := make(map[K]V)
+	oldInDPNotDesired := c.inDataplaneNotDesired
+	newInDPNotDesired := make(map[K]V)
+
+	err := iter(func(k K, v V) {
+		// Figure out if we _want_ it to exist and tee up update/deletion accordingly.
+		if desiredV, desired := c.GetDesired(k); desired {
+			// Record that this key exists in the new copy of the cache.
+			newInDPDesired[k] = v
+
+			// Check if the value is correct,
+			if c.valuesEqual(desiredV, v) {
+				// Value in dataplane is correct, clean up any pending update.
+				delete(c.desiredUpdates, k)
+			} else {
+				// Value in dataplane is incorrect.  Queue up an update (if there isn't one already).
+				c.desiredUpdates[k] = desiredV
+			}
+		} else {
+			// We don't want this key, but it's in the dataplane.  Queue up deletion.
+			newInDPNotDesired[k] = v
+		}
+
+		// Remove the key from the old cache, we'll then scan the old cache to find keys that are missing.
+		delete(oldInDPDesired, k)
+		delete(oldInDPNotDesired, k)
+	})
+	if err != nil {
+		// We may have failed mid-iteration, if we just returned an error here, we'd leave our internal
+		// state broken because we've removed all the keys that we've seen from oldInDPDesired and
+		// oldInDPNotDesired, and updated c.desiredUpdates to match the new KV's from the iterator.
+		// Fix that up by applying the new keys to the old in-DP maps:
+		for k, v := range newInDPDesired {
+			oldInDPDesired[k] = v
+		}
+		for k, v := range newInDPNotDesired {
+			oldInDPNotDesired[k] = v
+		}
+		return fmt.Errorf("failed to iterate over dataplane state: %w", err)
+	}
+
+	// oldInDPDesired now only contains KVs that we _thought_ were in the dataplane but are now gone.
+	for k := range oldInDPDesired {
+		if desiredV, desired := c.GetDesired(k); desired {
+			// We want this key, but it's missing, queue up an add.
+			c.desiredUpdates[k] = desiredV
```

Comment:
In the error case, the old-in -DP maps contain:

* Any keys that are are not in the dataplane.
* Any keys that are in the dataplane but that we didn't reach before the error ocurred.

Meanwhile, the "new" maps only contain keys that _are_ in the dataplane and that we saw earlier in the iteration.

Since the "desired" state is tracked partially by the in-dataplane-desired map, if we _just_ bailed out then we may have deleted some desired keys from that map and they'd be 100% lost.  So user could have done `SetDesired(x, y)` and then do `GetDesired(x)` and get nil :skull_and_crossbones: .

To correct that, we could either work on a copy of the in-DP-desired map and take the occupancy hit, or we have to fix it up after error.  I think I got the fix-up right: each key we iterate over either goes in the new-desired or the new-not-desired map and it gets deleted from the old maps, so we put back all the keys that we took out.  Of course, since the resync failed, the state of the dataplane was only partially updated so a new resync will be needed to clean up the remaining keys.

Perhaps its slightly overkill to avoid copying the maps but my reason for doing it was that I want to use this code in the IP sets logic and _that_ code already does this kind of map-shuffling optimisation (albeit in a different way).  I didn't want to regress the occupancy there because I think we tried hard to keep it low.

Reply:
We're iterating over the desired keys that we _thought_ were in the dataplane before and that we did non see as part of the iteration.  That means the keys were unexpectedly deleted from the dataplane outside Felix and we need to put them back.  Hence moving them from desired-in-dataplane to desired-updates.

Reply:
got it!